contracts/serverless-v2/CommonChainExecutors.sol

```diff
@@ -280,45 +335,98 @@ contract CommonChainExecutors is
         }
     }
 
-    function _selectExecutors(
+    function _selectNodes(
         uint256 _noOfNodesToSelect
     ) internal view returns (address[] memory selectedNodes) {
         uint256 randomizer = uint256(keccak256(abi.encode(blockhash(block.number - 1), block.timestamp)));
         selectedNodes = _selectN(randomizer, _noOfNodesToSelect);
         // require(selectedNodes.length != 0, "NO_EXECUTOR_SELECTED");
     }
 
-    function updateOnSubmitOutput(
+    function _updateOnSubmitOutput(
         address _executorKey
-    ) external onlyJobsContract {
+    ) internal {
         // add back the node to the tree as now it can accept a new job
         if(executors[_executorKey].status && executors[_executorKey].activeJobs == executors[_executorKey].jobCapacity)
             _insert_unchecked(_executorKey, uint64(executors[_executorKey].stakeAmount));
 
         executors[_executorKey].activeJobs -= 1;
 
+        // if user has initiated unstake then release tokens only if no jobs are pending
+        if(executors[_executorKey].unstakeStatus && executors[_executorKey].activeJobs == 0) {
+            uint256 amount = executors[_executorKey].stakeAmount < executors[_executorKey].unstakeAmount ? executors[_executorKey].stakeAmount : executors[_executorKey].unstakeAmount;
+            executors[_executorKey].stakeAmount -= amount;
+            TOKEN.safeTransfer(_msgSender(), amount);
+            executors[_executorKey].unstakeAmount = 0;
+            // update in tree only if the user has not initiated deregistration
+            if(executors[_executorKey].status)
+                _update_unchecked(_executorKey, uint64(executors[_executorKey].stakeAmount));
+        }
+
         if(!executors[_executorKey].status && executors[_executorKey].activeJobs == 0) {
             delete executors[_executorKey];
             _revokeEnclaveKey(_executorKey);
         }
     }
 
-    function updateOnExecutionTimeoutSlash(
-        address _executorKey
-    ) external onlyJobsContract {
+    function _updateOnExecutionTimeoutSlash(
```

Comment:
I have added `bool _hasExecutedJob` param in the `_updateOnExecutionTimeoutSlash()` function as it will be used for slashing. The logic before and after slashing is same so I'm creating 2 internal functions to handle them. Let me know if that looks good.